src/Http/RequestFactory.php

```diff
@@ -60,25 +60,59 @@ public function setProxy($proxy)
 	 */
 	public function createHttpRequest()
 	{
+		$remoteAddr = isset($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : NULL;
+		$remoteHost = isset($_SERVER['REMOTE_HOST']) ? $_SERVER['REMOTE_HOST'] : NULL;
+
+		// use real client address and host if trusted proxy is used
+		$usingTrustedProxy = FALSE;
+		foreach ($this->proxies as $proxy) {
+			if (Helpers::ipMatch($remoteAddr, $proxy)) {
+				$usingTrustedProxy = TRUE;
+				break;
+			}
+		}
+
+		if ($usingTrustedProxy) {
+			if (isset($_SERVER['HTTP_X_FORWARDED_FOR'])) {
+				$remoteAddr = trim(current(explode(',', $_SERVER['HTTP_X_FORWARDED_FOR'])));
+			}
+
+			if (isset($_SERVER['HTTP_X_FORWARDED_HOST'])) {
+				$remoteHost = trim(current(explode(',', $_SERVER['HTTP_X_FORWARDED_HOST'])));
+			}
+		}
+
 		// DETECTS URI, base path and script path of the request.
 		$url = new UrlScript;
-		$url->setScheme(!empty($_SERVER['HTTPS']) && strcasecmp($_SERVER['HTTPS'], 'off') ? 'https' : 'http');
+		$url->setScheme($scheme = $this->getScheme($usingTrustedProxy));
 		$url->setUser(isset($_SERVER['PHP_AUTH_USER']) ? $_SERVER['PHP_AUTH_USER'] : '');
 		$url->setPassword(isset($_SERVER['PHP_AUTH_PW']) ? $_SERVER['PHP_AUTH_PW'] : '');
 
-		// host & port
+		// host and port
+		$port = null;
 		if ((isset($_SERVER[$tmp = 'HTTP_HOST']) || isset($_SERVER[$tmp = 'SERVER_NAME']))
 			&& preg_match('#^([a-z0-9_.-]+|\[[a-f0-9:]+\])(:\d+)?\z#i', $_SERVER[$tmp], $pair)
 		) {
 			$url->setHost(strtolower($pair[1]));
 			if (isset($pair[2])) {
-				$url->setPort((int) substr($pair[2], 1));
+				$port = (int) substr($pair[2], 1);
 			} elseif (isset($_SERVER['SERVER_PORT'])) {
-				$url->setPort((int) $_SERVER['SERVER_PORT']);
+				$port = (int) $_SERVER['SERVER_PORT'];
 			}
+		} else if (!empty($_SERVER['SERVER_PORT'])) {
+			$port = (int) $_SERVER['SERVER_PORT'];
+		}
+
+		if ($usingTrustedProxy && !empty($_SERVER['HTTP_X_FORWARDED_PORT'])) {
+			$port = (int) $_SERVER['HTTP_X_FORWARDED_PORT'];
+		}
+
+		$isInvalidCombination = $scheme === 'http' && $port === 443 || $scheme === 'https' && $port === 80;
```

Comment:
I omitted them intentionally - it works well. Anyway if we don't test it there then we can endup with strange urls like `https://localhost:80`. Of course we can decide that this is ok and it's up to the developer to take care of these invalid links. The server would have to be badly missconfigured to produce such a strange combination of scheme and port.